### PR TITLE
Developer's Theory Guide for Website

### PR DESCRIPTION
**Pull request by kkiesling**
The PR is for the developer's theory guide for the website. It would be appreciated if people could review the content for correctness.

### CODE REVIEW

File: doc/contribute/devtheory.rst [hunk 1]
```diff
@@ -0,0 +1,185 @@
+Developer's Theory Guide
+========================
+
+This guide is meant to give developers a comprehensive understanding of how
+DAGMC works as it exists in MOAB and the key components one may use during
+development.
+
+Contructor
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: Spelling

File: doc/contribute/devtheory.rst [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+~~~~~~~~~~~~
+
+To initialize a DAGMC instance, the constructor must be called. If the
+constructor is called with no arguments, then a new instance is created. Else,
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: Replace "Else" w/ "Otherwise" perhaps?

Comment (pshriwise) [reply doc/contribute/devtheory.rst]: There's a need to be specific about what instances are being talked about here. (MOAB instance vs DAGMC instance)

File: doc/contribute/devtheory.rst [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+There are three main steps to setting up a DAGMC geometry. First the storage
+space for the implicit complement is created. Then the OBB trees are built.
+And finally the sets of indices for volumes and surfaces are generated.
+See sections below for more information about implicit complements, OBBs,
+and indices.
+
+Topology
+--------
+
+Every geometry contains entities that are either volumes, surfaces, and curves.
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: replace "and" w/ "or" here

File: doc/contribute/devtheory.rst [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+DAGMC Setup
+~~~~~~~~~~~~
+
+There are three main steps to setting up a DAGMC geometry. First the storage
+space for the implicit complement is created. Then the OBB trees are built.
+And finally the sets of indices for volumes and surfaces are generated.
+See sections below for more information about implicit complements, OBBs,
+and indices.
+
+Topology
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst:26]: This is probably the most difficult concept to convey and might merit some kind of diagram or tree.
The difference between these geometry entities represented as MOAB entity sets with parent-child relationships to these sets which in turn have a containment relationship with mesh entities that replace the analytic geometric representation. With that in mind I'd be really specific when using the word entity, including some clarifying term in front.

File: doc/contribute/devtheory.rst [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+however, is empty. While a volume is made of surfaces (the parent-child
+relationship), the volume does not contain an entity set.
+
+Sense Tags
+----------
+
+Each surface is tagged with the two volume handles of the adjacent volumes.
+The first of the two surfaces is designated as the forward direction and the
+second is designated with the reverse direction. It is important to note that
+these surfaces senses may not be consistent with how an MC code determines
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst:48]: Well said in the last sentence.

File: doc/contribute/devtheory.rst [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+necessarily have to be contiguous. During setup, the global ID is tagged on the
+entity set. The second method is to identify by index. This is an ordinal
+numerical reference for each volume or surface in the geometry. The third method
+is to identify by the entity handle of the entity set. The entity handle is
+assigned by MOAB and is not necessarily contiguous.
+
+Cross Referencing
+-----------------
+
+If one type of indice is known, then either of the other two can be determined.
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: index rather than indice I believe.

File: doc/contribute/devtheory.rst [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+By using the functions below or a combination of the functions, either ID,
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: Global ID for clarity?

File: doc/contribute/devtheory.rst [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+handle (index_by_handle).
+
+OBB Tree
+~~~~~~~~
+
+What is an OBB?
+---------------
+
+An OBB, or Oriented Bounding Box, is a box that contains a single facet or set
+of facets and has an axis oriented such that it is the smallest possible box
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: all of the box axes are technically oriented

File: doc/contribute/devtheory.rst [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    :width:  300
+    :alt:    Image of a 3-D Oriented Bounding Box (OBB) around a facet
+
+OBB Tree Construction
+---------------------
+
+First, the OBB tree for each surface is built. This is built top down where the
+top level is an OBB that contains all the surface’s facets. Then the set of facets
+is split roughly in half and two new OBBs form around each of the new sets.
+This continues until each of the OBBs the level contains a single facet. A 2D
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: I think the default value for number of leaf node facets is 8

Comment (pshriwise) [reply doc/contribute/devtheory.rst]: Not necessary to alter the existing image though. I'd just be vague and indicate that it contains a small number of facets.

File: doc/contribute/devtheory.rst [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+Ray History
+~~~~~~~~~~~
+
+A particle can be either streaming or reflecting. If streaming, the particle
+may only cross a particular surface once. To ensure this a list (or history)
+of the surfaces that particle has crossed is maintained. If a surface encountered
+is reflecting, the particle does not cross that surface and is therefore not
+added to the history. If a particle is or becomes a reflecting particle, the
+history of crossed surfaces is cleared.
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: I think there is more to this construct. @makeclean knows it the best.

File: doc/contribute/devtheory.rst [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+~~~~~~~~~~~~~~~
+
+Given a volume entity handle, position, and ray direction (optional), the
+point_in_volume function will test if the point is inside or outside the given
+volume. It is assumed that the test volume exists and is known. Passing a
+direction vector to this function adds robustness and ensures consistent results.
+Otherwise, a random direction is used.
+
+Ray Fire
+~~~~~~~~
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst:188]: It is my understanding that this function can only be called for one volume at a time. The bit about ray history here is spot on though.

File: doc/contribute/devtheory.rst [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+than once (ie, if the particle is streaming).
+
+Next Volume
+~~~~~~~~~~~
+
+If the next surface is known (after calling ray_fire), the entity handle of the
+next volume can be determined by calling next_volume. Given the next surface and
+the known current volume, the next volume is determined by looking at the other
+volume tagged on that surface (as described in the Sense Tags section above).
+This assumes that a valid surface and volume are provided. If no next volumes
```
Comment (pshriwise) [inline doc/contribute/devtheory.rst]: volumes to volume

**Review by pshriwise — COMMENTED** [review]
Mostly minor changes for clarity or typos. This will be very valuable to people trying to understand the code for the first time. Thanks!

**Comment by ljacobson64** [comment]
Just in terms of style, I think it would be helpful if names of functions were formatted as code; i.e. `point_in_volume` instead of point_in_volume

File: doc/contribute/devtheory.rst (hunk 2 above)
Comment (kkiesling) [reply doc/contribute/devtheory.rst]: Okay, I think I made the proper clarification, but I am unsure about what exactly is happening. Let me know if how I wrote it up now is correct.

File: doc/contribute/devtheory.rst (hunk 7 above)
Comment (kkiesling) [reply doc/contribute/devtheory.rst]: yes, thanks for catching that

File: doc/contribute/devtheory.rst (hunk 9 above)
Comment (kkiesling) [reply doc/contribute/devtheory.rst]: Thanks for clarifying that.

File: doc/contribute/devtheory.rst (hunk 11 above)
Comment (kkiesling) [reply doc/contribute/devtheory.rst:188]: yes that is correct about only being able to call it one volume at a time and probably a good thing to clarify or note here

File: doc/contribute/devtheory.rst (hunk 8 above)
Comment (kkiesling) [reply doc/contribute/devtheory.rst]: I think I correctly clarified what you are talking about now, but can you review the new statement and let me know if it is or is not correct now.

File: doc/contribute/devtheory.rst (hunk 4 above)
Comment (kkiesling) [reply doc/contribute/devtheory.rst:26]: Let's clarify this in person, I am not sure what the best type of diagram is for this clarification.

**Comment by kkiesling** [comment]
Thanks so much @Pshriwise and @ljacobson64 for the initial review! I believe I have addressed all of your comments except for clarifying the topology section. I think I may need a bit of discussion in person to determine the best type of diagram to put there. @Pshriwise if you could just check that my changes to the constructor section and OBB section correctly take into account your comments. Thanks again.

File: doc/contribute/devtheory.rst [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+Types of Indices
+----------------
+
+There are three ways to identify any surface and volume in a geometry:
+global ID, index, or entity handle. The global ID is an integer number that
+corresponds to the Cubit ID. As a result, the set of global IDs does not
+necessarily have to be contiguous. During setup, the global ID is tagged on the
+entity set. The second method is to identify by index. This is an ordinal
+numerical reference for each volume or surface in the geometry. The third method
```
Comment (kkiesling) [inline doc/contribute/devtheory.rst]: Index starts from 1

**Comment by gonuke** [comment]
I think this only needs an update on RayHistory from @makeclean and then it will be ready..

**Comment by ljacobson64** [comment]
It will also need to have the base branch changed to develop and be rebased against develop as well

**Comment by ljacobson64** [comment]
Thanks @kkiesling!

**Pull request merged**